analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/HashesShouldHaveUnpredictableSaltBase.cs

```diff
@@ -18,12 +18,86 @@
+
+    private const int SafeSaltSize = 16;
+    private const string MakeSaltUnpredictableMessage = "Make this salt unpredictable.";
+    private const string MakeThisSaltLongerMessage = "Make this salt at least 16 bytes.";
+
+    private static readonly ImmutableArray<MemberDescriptor> CryptographicallyStrongRandomNumberGenerators = ImmutableArray.Create(
+        new MemberDescriptor(KnownType.System_Security_Cryptography_RandomNumberGenerator, nameof(RandomNumberGenerator.GetBytes)),
+        new MemberDescriptor(KnownType.System_Security_Cryptography_RandomNumberGenerator, nameof(RandomNumberGenerator.GetNonZeroBytes)));
+
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        var state = context.State;
+        var instance = context.Operation.Instance;
+        if (instance.AsObjectCreation() is { } objectCreation)
+        {
+            return ProcessObjectCreation(objectCreation, state);
+        }
+        else if (instance.AsInvocation() is { } invocation)
+        {
+            return ProcessInvocation(invocation, state);
+        }
+        else if (instance.AsArrayCreation() is { } arrayCreation)
+        {
+            return ProcessArrayCreation(arrayCreation, state);
+        }
+        return state;
+    }
+
+    private ProgramState ProcessObjectCreation(IObjectCreationOperationWrapper objectCreation, ProgramState state)
+    {
+        if (objectCreation.Type.DerivesFrom(KnownType.System_Security_Cryptography_DeriveBytes)
+            && objectCreation.Arguments.FirstOrDefault(x => x.AsArgument() is { Parameter.Name: "salt" or "rgbSalt" }) is { } saltArgument)
+        {
+            if (state[saltArgument]?.HasConstraint(ByteCollectionConstraint.CryptographicallyWeak) is true)
+            {
+                ReportIssue(saltArgument, MakeSaltUnpredictableMessage);
+            }
+            else if (state[saltArgument]?.HasConstraint(SaltSizeConstraint.Short) is true)
+            {
+                ReportIssue(saltArgument, MakeThisSaltLongerMessage);
+            }
+        }
+        return state;
+    }
+
+    private static ProgramState ProcessInvocation(IInvocationOperationWrapper invocation, ProgramState state)
+    {
+        if (CryptographicallyStrongRandomNumberGenerators.Any(x =>
+                invocation.TargetMethod.Name == x.Name
+                && invocation.TargetMethod.ContainingType.DerivesFrom(x.ContainingType))
+            && invocation.ArgumentValue("data") is { } dataArgument
+            && dataArgument.TrackedSymbol() is { } trackedSymbol)
+        {
+            state = state.SetSymbolConstraint(trackedSymbol, ByteCollectionConstraint.CryptographicallyStrong);
+        }
+        return state;
+    }
+
+    private static ProgramState ProcessArrayCreation(IArrayCreationOperationWrapper arrayCreation, ProgramState state)
+    {
+        if (arrayCreation.Type.Is(KnownType.System_Byte_Array)
```

Comment:
I saw these UTs in the rule I'm working in: https://github.com/SonarSource/sonar-dotnet/blob/e8a0d4a4c1f61d605c49b5ed787e593d774bd3f6/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Sonar/InitializationVectorShouldBeRandom.cs#L230C20-L230C41

This array seems to be considered cryptographically strong and I think we can detect it.
Is it the case also here? Maybe we should contact appsec.

Reply:
As agreed in an offline discussion, we'll contact AppSec and address this issue in a potential future PR.